src/Database/TigerBeetle/Raw/Client.hs

```diff
@@ -146,3 +145,110 @@ withClient cfg clusterId address action =
         (fromIntegral addressLen)
         0
         callback
+
+      case initStatus of
+        Success -> finally
+            (Right <$> action clientState)
+            (finalizeClient clientState)
+        _ -> pure $ Left initStatus
+
+finalizeClient :: ClientState -> IO ()
+finalizeClient state = do
+  -- Signal that no new incoming request should proceed
+  atomically $ writeTVar state.csIsShutdown True
+
+  -- N.B. race-condition: Once we release from this transaction
+  -- completionCallbacks may fire and overwrite our shutdown error in the
+  -- request resultVar. Conceivably this is fine, the below block just ensures
+  -- that all submitRequest invocations that are blocked on the request context
+  -- resultVar are allowed to proceed. This should effectively flush all active
+  -- requests since the isShutdown flag should prevent new requests from coming
+  -- in.
+  --
+  -- Process interleaving shouldn't cause any active request entries after
+  -- this block (due to using a stale isShutdown value) because the flag is
+  -- read in the same atomically block as active requests are written to in
+  -- submitRequest function
+  atomically $ do
+    activeReqs <- readTVar state.csActiveRequests 
+    -- Iterate through all the request vars and put a client shutdown result 
+    -- so that instances of `submitRequest` that are waiting are unblocked
+    forM_ (IM.elems activeReqs) $ \context ->
+      putTMVar context.resultVar (Left ClientShutdownDuringRequest)
+  
+  -- De-initialize the client
+  void $ tbClientDeinit state.csClientPtr
+
+submitRequest 
+  :: ClientState
+  -> TBOperation
+  -> ByteString  -- ^ Request data
```

Comment:
I think we could come up with a higher level `submitRequest` function that operates on a single sum type where the constructor indicates the `TBOperation` and contains a payload that can be converted to the correct `TB*` entity which can then be serialized into this bytestring.